File: Flyweight/FlyweightGame/FlyweightGame/SoldierFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FlyweightGame
{
    public class SoldierFactory
    {
        private Dictionary<string, SoldierType> _types = new();

        public SoldierType GetSoldierType(string type, string color, string animation)
        {
            string key = $"{type}_{color}";

            if (!_types.ContainsKey(key))
            {
                _types[key] = new SoldierType(type, color, animation);
            }

            return _types[key]; 
        }

        public int TotalTypes => _types.Count;
    }
}

File: Flyweight/JavaExample/FlyweightGame/src/SoldierFactory.java
import java.util.HashMap;
import java.util.Map;

public class SoldierFactory {
    private final Map<String, SoldierType> types = new HashMap<>();

    public SoldierType getSoldierType(String type, String color, String animation) {
        String key = type + "_" + color;

        if (!types.containsKey(key)) {
            types.put(key, new SoldierType(type, color, animation));
        }
        return types.get(key);
    }

    public int getTotalTypes() {
        return types.size();
    }
}
